### DISCUSSION TITLE
Add table of contents to documentation pages

### DISCUSSION THREAD

**Discussion by billytetrud** [sqlalchemy/sqlalchemy#8295] [Ideas]
Eg, how am I supposed to find anything in this 30 page document? https://docs.sqlalchemy.org/en/14/orm/query.html Please have an index where I can easily find a list of methods.

**Comment by zzzeek** [comment]
hey there -

[sphinx](https://www.sphinx-doc.org/en/master/) offers what they offer here, which is the [autodoc](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#module-sphinx.ext.autodoc) extension.  it does not make a TOC of the actual methods which I assume is what you are complaining about.

As far a Query, first off, [it's legacy](https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#orm-query-unified-with-core-select), secondly, the main documentation for "how to do different query patterns" is in the [orm query guide](https://docs.sqlalchemy.org/en/14/orm/queryguide.html) which is more in terms of the newer patterns.  

finally, if I'm on https://docs.sqlalchemy.org/en/14/orm/query.html and really have to find something, I use the "ctrl-f" feature of the browser just to find the name, or the search box in the upper left which works great for specific method names, like typing [query.filter_by](https://docs.sqlalchemy.org/en/14/search.html?q=query.filter_by&check_keywords=yes&area=default) will take you right there.

**Reply by billytetrud** [reply]
Thanks for the notes. I am of course aware of ctrl-f.. My primary complaint is that its hard to get a quick overview of what functionality exists without painstakingly reading the entire documentation from cover to cover. These docs could be helped by splitting up [the various types of documentation](https://documentation.divio.com/reference/). In particular, there is overuse of examples in the docs that should be split out into a separate set of documents that teach how to use alchemy.

**Reply by zzzeek** [reply]
hi - 

the use case of "finding something specific" and getting an "overview of what functionality exists" are opposites.

For the "overview of what functionality exists",  please spend a little more time with the docs, in particular looking over [the index page](https://docs.sqlalchemy.org/en/14/), where you'll see things like the Quickstart and the Tutorial, which is where people should be starting when they want to get an overview of SQLAlchemy functionality.   those are the **narrative** docs that will explain the concepts you need to know as well as bring awareness to the key functions and methods that are important to know about.   

Also I don't have much interest in "Divio" but as it's a commercial product it looks like they had a dedicated, salaried team writing their documentation.  good for them!  we're an open source project.  it's me and two or three other volunteers right now.    Please keep that in mind when using OSS software.

**Reply by billytetrud** [reply]
> the use case of "finding something specific" and getting an "overview of what functionality exists" are opposites.

They're not when you don't know if you don't know if the specific thing you're looking for exists or what its name might be.

> when they want to get an overview of SQLAlchemy functionality

That is not what I'm looking for.

> those are the narrative docs

Again, I'm looking for reference docs, not narrative docs (which I think maps closer to divio's concept of "explanation docs").

> Also I don't have much interest in "Divio"

[You asked how you can improve your docs](https://twitter.com/zzzeek/status/1403361145357062144). I'm referencing a very well thought out documentation strategy. It doesn't really sound like you want advice tho, so I guess you can take it or leave it.

**Reply by CaselIT** [reply]
> It doesn't really sound like you want advice tho, so I guess you can take it or leave it.

I think the advice should be targeted to something actionable.
Like is there a way to make sphinx output the method index for a class like query?

**Reply by zzzeek** [reply]
I have  a ton of interest in improving the documentation and I work on them all the time, dozens of hours in the past few weeks in fact.    However I don't have the time or mental capacity to redo the entirety of the documentation to suit some particular structure, which seems to have at the very least a ton of overlap with what we have anyway.     This is not something I can do, I have been writing these docs for 16 years, so suggestions like "read this book on how to write docs and then rewrite your whole site to conform to it exactly" is not a thing that is actionable on this end by me personally - there's no way I can by myself make that jump intellectually at this point - for commercial organizations like Divio (or as I see they mention Django, gigantic open source projects with hundreds of highly competent volunteers and a foundation putting money directly into it, not to mention a very different kind of product that is oriented towards beginners in the first place), they actually don't have the developers write the docs at all, because we can't - we're too deep into it, they have paid teams of people who can bring better perspectives.      We don't have that.   So "read this website and get to work" is not going to happen for me.  That's not a constructive criticism and suggests you haven't made a deep investment in at least understanding the problem our docs are trying to solve.    

Constructive criticism would come from a place where you've spent many hours / days / weeks with our docs to understand the problem space very deeply and you can then offer specific organization approaches that are actionable, such as a Sphinx extension that produces tables of contents for very long class docs.    That's of course a lot of work.   Re: the tweet, I was asking people to show specific sections where they read the paragraphs and they didnt understand what they were reading, i.e., specifics.    The specific complaint of "Query doesn't have a table of contents" I dont really know what that's asking for, that page is a docstring for a single Python class, it's not where one would be learning Query from.

**Reply by zzzeek** [reply]
going through the Divio guide:

1. our Tutorials are written in the style that they describe , and the latest ones are [here](https://docs.sqlalchemy.org/en/14/orm/quickstart.html) and [here](https://docs.sqlalchemy.org/en/14/tutorial/index.html).  they most certainly follow all the guidelines at https://documentation.divio.com/tutorials/ , very intentionally and explicitly.    They are award-winning examples of those principles.   A profound amount of effort has been put into these tutorials because that's where the beginners are.  
2. our Reference is most everything else in the docs.  
3. we don't have "how-to" guides for the most part.   Reference is so gigantic as it is, within reference there will be small examples of things, but this isn't a "how-to" guide.    Not all docs have "how-to" guides, for example the documentation for [postgresql](https://www.postgresql.org/docs/current/index.html) and [mysql](https://dev.mysql.com/doc/).   This certainly is very much something I can see Django having, though.     we have an [examples](https://docs.sqlalchemy.org/en/14/orm/examples.html) section and a [wiki recipes](https://github.com/sqlalchemy/sqlalchemy/wiki/UsageRecipes) section, which are more or less "how to" but these aren't as polished as I'd like them to be.
4. "explanation" - we dont have much of this at all as in my experience, nobody reads it and the people who do read it get generally overwhelmed and confused.    We do have some "explanatory" elements that are very intentionally kept "separate" as inline sidebars.  they are sidebars first so that users know they don't have to read them to proceed with the docs they are reading, and also as sidebars they are kept very short and minimal.   someone who really wants to do a deep dive can go to [the SQLAlchemy library](https://www.sqlalchemy.org/library.html) and either read our AOSA chapter or watch talks.

This is why the link to these Divio docs rubbed me the wrong way, because we are already doing all these things.     They just are doing them "better" because they have paid staff writing their docs.

**Reply by billytetrud** [reply]
I'm specifically making two actionable recommendations:
* Have an index of classes and methods so I can look through the names and click on links to them to read more. This would help me find obscure things I would otherwise have to scour the docs for over an hour.
* Separate the "explanation" part of the docs into literally separate pages, not just differently colored inline boxes. This would massively declutter the reference docs. 

Thanks

**Reply by zzzeek** [reply]
> I'm specifically making two actionable recommendations:
> 
>     * Have an index of classes and methods so I can look through the names and click on links to them to read more. This would help me find obscure things I would otherwise have to scour the docs for over an hour.

the search box as well as [the index](https://docs.sqlalchemy.org/en/14/genindex.html) has this.  IMO the index is way too much to be able to find things, I think you should navigate to the subsection of functionality that you want, then you'll see I in fact wrote a Sphinx extension for this purpose on each page as you can see [here](https://docs.sqlalchemy.org/en/14/orm/mapping_api.html).    your only example given is Query which is kind of an outlier here.   The sphinx extension can perhaps be given an option to work on a method basis for the few classes like Query that have lots of methods.

> 
>     * Separate the "explanation" part of the docs into literally separate pages, not just differently colored inline boxes. This would massively declutter the reference docs.

I dont have faith that you've even seen these boxes, since there are very few of them only in specific parts of the tutorial and almost never at all in the reference documentation, and nobody has a problem with them as they are very clear and succinct and only serve to enhance the docs.  I have seen only positive comments about them.. please show specific examples of "clutter" so I know what you are referring towards.

**Reply by zzzeek** [reply]
I've added javascript functionality to classes with long lists of methods so that the methods and attributes are collapsed by default, so that you can scroll down the list of methods with just one line descriptions.  this is the only way I can see "A table of contents" for the methods of classes to be possible.

https://docs.sqlalchemy.org/en/20/orm/query.html?highlight=query#sqlalchemy.orm.Query

https://docs.sqlalchemy.org/en/20/orm/mapping_api.html?highlight=mapper#sqlalchemy.orm.Mapper

it's not totally polished as of yet but this had to be written totally custom

**Reply by zzzeek** [reply]
nevermind, that idea has a few showstoppers in it.    will try something else

**Reply by CaselIT** [reply]
Personally I was not hating it. What did you not like?

**Reply by zzzeek** [reply]
well fixable was it was blowing away all the "parameter" boxes so it had to come down.

but showstopper without any real fix except for something extremely elaborate, it breaks all `:paramref:.Column.some_method.someparam` references, because they aren't shown

**Reply by zzzeek** [reply]
here's another idea i have here.   looks way too busy

![Screenshot from 2022-07-31 11-08-57](https://user-images.githubusercontent.com/128223/182032837-e43871d9-a449-4680-afc9-518cb398af57.png)

**Reply by zzzeek** [reply]
so let me table that and then try somethign w/ the "more" box since you liked it (which i am surprised by)

**Reply by CaselIT** [reply]
not really liked, not hated it.

Maybe a right hand side sidebar with the class method is better though. But it does not look to be easy especially if there are multiple classes in the same doc

**Reply by zzzeek** [reply]
so i can get the "more" box to open up if you navigate to the page w/ an anchor, but it's not lining up great.  anchors *within* the page I'd have to make another event.  like the "more" box thing is a LOT of jquery.    feels too risky

**Reply by zzzeek** [reply]
yet another idea is the "list all the methods" thing above and have *that* inside of a "more" box.  but I dont like that either
